### ISSUE TITLE
"grad" is "None" on sliced section of Variable

### ISSUE THREAD

**Issue by benvcutilli** [pytorch/pytorch#4649]
If I had something like:

```
superVariable = torch.autograd.Variable(torch.ones(4,4))
subVariable = superVariable[1:3].cuda()
network.forward(subVariable).backward()
print(subVariable.grad)
```

`subVariable.grad` will show as `None`, but `superVariable.grad` is not `None`. Not sure if this is a bug, a feature request, or if I am doing something wrong. I've checked that `subVariable.requires_grad` is `True`. This would be incredibly convenient in the case of if we have a lower-case "v" variable already dedicated to a mini batch, and we want the gradient w.r.t that mini batch.

Sorry if I missed something important. I took a semi-close look at the documentation for any obvious explanation, but came up with nothing. Also, `torch.__version__` says `'0.3.0.post4'`.

**Comment by apaszke** [comment]
Gradient is accumulated only in leaf Variables (i.e. parameters/created by the user). You can call `.retain_grad()` on `subVariable` if you want it to get `.grad` saved.

Note that we're using GitHub issues for bug reports only, and all questions should be posted on [our forums](discuss.pytorch.org).

**Comment by benvcutilli** [comment]
Ok, thanks. Like I said, this could have very well been a feature request too, and I am aware of the forums.

**Comment by benvcutilli** [comment]
I would also mention that `retain_grad()`'s description in the documentation says `non-leaf Variables`, which it is not clear what is and is not a leaf Variable when many, including myself, would think that when you slice a leaf Variable, you are still working with a Variable that is a leaf in the sense that it is the same data, just a subsection of it. It may be a new lower-case "v" variable, but isn't obviously a new upper-case "V" Variable.

Just a suggestion that would help to clarify usage.

**Comment by apaszke** [comment]
I'm pretty sure there are a few places where we define the term "leaf Variable", but I agree it would be a good idea to always link to them from the docs. Could you please send a PR?

**Comment by benvcutilli** [comment]
Well I mean I can try, but I would rather see a change where Variables sliced from others keep the gradients for each of the slice's elements, even if it is sliced before the backwards. Intuitively this makes sense; slicing alone doesn't change the equation that you are differentiating, so the gradient wouldn't change, and in normal Python, when slicing a list of objects, you end up with not just the objects but all their fields too. Further, there'd be little memory increase (unless for some reason the overhead is a lot larger than I'd think it would be) as you'd just point back to the respective slice in the `.grad` field of `superVariable`.

So I guess consider this as a feature request. There may have already been a discussion on this, but if not, that's my recommendation.

**Comment by benvcutilli** [comment]
Also, I would just add that the linguistic distinction between `requires_grad` and `retain_grad` is a little confusing (in fact, the confusion partially caused me to file this issue in the first place, I suppose). Maybe `requires_grad` should be something like `grad_intermediary` or something way more melodic, word-wise, to communicate the point that it will have a gradient flowing through it, but it will not actually keep it stored after the gradient has flowed.
